core/src/main/java/org/elasticsearch/index/IndexSettings.java

```diff
@@ -129,6 +129,15 @@
     public static final Setting<Integer> MAX_ANALYZED_OFFSET_SETTING =
         Setting.intSetting("index.highlight.max_analyzed_offset", 10000, 1, Property.Dynamic, Property.IndexScope);
 
+
+    /**
+     * Index setting describing the maximum number of terms that can be used in Terms Query.
+     * The default maximum of 2000 terms is defensive, as extra processing and memory is involved
+     * for each additional term, and a large number of terms degrade the cluster performance.
+     */
+    public static final Setting<Integer> MAX_TERMS_COUNT_SETTING =
+        Setting.intSetting("index.max_terms_count", 2000, 1, Property.Dynamic, Property.IndexScope);
+
```

Comment:
Originally, I intended to set the limit to `1024` the same number as the Lucene maximum number of boolean clauses, but some tests in xpack use higher number of terms, thus  I had to increase this number to 2000.